actors.md

```diff
@@ -478,6 +478,185 @@ func UpdatePeerID(pid PeerID) {
+
+#### State
+
+```go
+type Multisig struct {
+    Signers []Address
+    Required uint
+    
+    NextTxID uint64
+    Transactions map[int]Transaction
+}
+
+type Transaction struct {
+    TxID uint64
+    To Address
+    Value TokenAmount
+    Method string
+    Params []byte
+    Approved []Address
+    Completed bool
+    Canceled bool
+}
+```
+
+
+
+#### Constructor
+
+```go
+func Multisig(signers []Address, required uint) {
+    self.Signers = signers
+    self.Required = required
+}
+```
+
+
+
+### Propose
+
+```go
+func Propose(to Address, value TokenAmount, method string, params []byte) uint64 {
+    if !isSigner(msg.From) {
+        Fatal("not authorized")
+    }
+    
+    txid := self.NextTxID
+    self.NextTxID++
+    
+    tx := Transaction{
+        TxID: txid,
+        To: to,
+        Value: value,
+        Method: method,
+        Params: params,
+        Approved: []Address{msg.From},
+    }
+    
+    self.Transactions.Append(tx)
+    
+    return txid
+}
+```
+
+### Approve
+
+```go
+func Approve(txid uint64) {
+    if !isSigner(msg.From) {
+        Fatal("not authorized")
+    }
+    
+    tx := getTransction(txid)
+    if tx.Complete {
+        Fatal("transaction already completed")
+    }
+    if tx.Canceled {
+        Fatal("transaction canceled")
+    }
+    
+    for _, signer := range tx.Approved {
+        if signer == msg.From {
+            Fatal("already signed this message")
+        }
+    }
+    
+    tx.Approved.Append(msg.From)
+    
+    if len(tx.Approved) > self.Required {
+        Send(tx.To, tx.Value, tx.Method, tx.Params)
+        tx.Complete = true
+    }
+}
+```
+
+### Cancel
+
+```go
+func Cancel(txid uint64) {
+    if !isSigner(msg.From) {
+        Fatal("not authorized")
+    }
+    
+    tx := getTransaction(txid)
+    if tx.Complete {
+        Fatal("cannot cancel completed transaction")
+    }
+    if tx.Canceled {
+        Fatal("transaction already canceled")
+    }
+    
+    tx.Canceled = true
+}
+```
+
+### ClearCompleted
+
+```go
+func ClearCompleted() {
+    if !isOwner(msg.From) {
+        Fatal("not authorized")
+    }
+    
+    for tx in range self.Transactions {
+        if tx.Completed || tx.Canceled {
+            self.Transactions.Remove(tx)
+        }
+    }
+}
+```
+
+
+
+### AddSigner
+
+```go
+func AddSigner(signer Address) {
+    if msg.From != self.Address {
+        Fatal("add signer must be called by wallet itself")
+    }
+    if isSigner(signer) {
+        Fatal("new address is already an owner")
+    }
+    
+    self.Signers.Append(signer)
+}
+```
+
+### RemoveSigner
+
+```go
+func RemoveSigner(signer Address) {
+    if msg.From != self.Address {
+        Fatal("remove signer must be called by wallet itself")
+    }
+    if !isSigner(signer) {
+        Fatal("given address was not a signer")
+    }
+    
+    self.Signers.Remove(signer)
+}
+```
+
+### ChangeRequirement
+
+```go
+func ChangeRequirement(req int) {
+    if msg.From != self.Address {
```

Comment:
how should this be called? is Address the general address of an actor?

Reply:
The way this works is you tell the wallet to send a multisig transaction to itself calling this method. that way we don't have to implement the signature threshold stuff differently for each of these changes.

Reply:
from reading the PR notes (not obvious in code itself) i am guessing the self.address is a well-signed threshold, ie m-of-n signers.